omp/distributed/index_map_kernels.cpp

```diff
@@ -0,0 +1,241 @@
+
+    // precompute the range id and part id of each input element
+    vector<size_type> range_ids(input_size, exec);
+    vector<comm_index_type> full_remote_part_ids(input_size, exec);
+    size_type range_id = 0;
+#pragma omp parallel for firstprivate(range_id)
+    for (size_type i = 0; i < input_size; ++i) {
+        range_id = find_range(recv_connections_ptr[i], part, range_id);
+        range_ids[i] = range_id;
+        full_remote_part_ids[i] = part_ids[range_ids[i]];
+    }
+
+    // sort by part-id and recv_connection
+    auto sort_it = detail::make_zip_iterator(
+        full_remote_part_ids.begin(), recv_connections_ptr, range_ids.begin());
+    std::sort(sort_it, sort_it + input_size, [](const auto& a, const auto& b) {
+        return std::tie(std::get<0>(a), std::get<1>(a)) <
+               std::tie(std::get<0>(b), std::get<1>(b));
+    });
+
+    // get only unique connections
+    auto unique_end = std::unique(
+        sort_it, sort_it + input_size, [](const auto& a, const auto& b) {
+            return std::tie(std::get<0>(a), std::get<1>(a)) ==
+                   std::tie(std::get<0>(b), std::get<1>(b));
+        });
+    auto unique_size = std::distance(sort_it, unique_end);
+
+    remote_global_idxs.resize_and_reset(unique_size);
+    auto remote_global_idxs_ptr = remote_global_idxs.get_data();
+    remote_local_idxs.resize_and_reset(unique_size);
+    auto remote_local_idxs_ptr = remote_local_idxs.get_data();
+
+    // store unique connections, also map global indices to local
+#pragma omp parallel for
+    for (size_type i = 0; i < unique_size; ++i) {
+        remote_global_idxs_ptr[i] = recv_connections_ptr[i];
+        remote_local_idxs_ptr[i] =
+            map_to_local(recv_connections_ptr[i], part, range_ids[i]);
+    }
+
+    // compute number of connections per part-id
+    vector<unsigned long long int> full_remote_sizes(part->get_num_parts(), 0,
+                                                     exec);
+
+#pragma omp parallel for
+    for (size_type i = 0; i < unique_size; ++i) {
+#pragma omp atomic
+        full_remote_sizes[full_remote_part_ids[i]]++;
+    }
+    auto num_neighbors =
+        full_remote_sizes.size() -
+        std::count(full_remote_sizes.begin(), full_remote_sizes.end(), 0);
+
+    remote_sizes.resize(num_neighbors);
+    remote_part_ids.resize_and_reset(num_neighbors);
+    {
+        size_type idx = 0;
+        for (size_type i = 0; i < full_remote_sizes.size(); ++i) {
+            if (full_remote_sizes[i] > 0) {
+                remote_part_ids.get_data()[idx] =
+                    static_cast<comm_index_type>(i);
+                remote_sizes[idx] = static_cast<int64>(full_remote_sizes[i]);
+                idx++;
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
+    GKO_DECLARE_INDEX_MAP_BUILD_MAPPING);
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void map_to_local(
```

Comment:
Even though these are kernel implementation, a short documentation would make it a bit easier to understand what these kernels are supposed to do.

Reply:
I've put some documentation into `index_map_kernels.hpp`. I think it's the most central place for that, but due to the macros it will not work with doxygen. But I think this is not a big issue, since I would expect that people looking for this kind of documentation would look into the code directly anyway.